### PR TITLE
New PyTorch DataLoader implementation using batched operations

### PR DESCRIPTION
**Pull request by fps7806**
The current DataLoader implementation relies on python iteration over individual rows of a parquet dataset. This is reasonable for small datasets but runs into throughput limitations:

Previous timing:
```
Samples per second for batch 10: 7.556e+04
Samples per second for batch 100: 1.001e+05
Samples per second for batch 1000: 1.083e+05
Samples per second for batch 100000: 4.364e+04
```

The new approach converts data into torch format BEFORE appending to the shuffling buffer and then relies on batched operations to perform all data appending/sampling.

New timing:
```
Samples per second for batch 10: 4.363e+05
Samples per second for batch 100: 1.468e+06
Samples per second for batch 1000: 1.607e+06
Samples per second for batch 100000: 1.141e+06
```
New timing with GPU-buffering:
```
Samples per second for batch 10: 3.87e+05
Samples per second for batch 100: 1.95e+06
Samples per second for batch 1000: 2.307e+06
Samples per second for batch 100000: 5.053e+06
```

In the best-case scenario, the samples/s increases 2 orders of magnitude. Although image-based problems would likely not achieve this throughput, these numbers are more representative of the throughput on tabular-data problems.

The proposed changes can be used out-of-the-box in most scenarios.

There are still many improvements that can be made to the ShufflingBuffer logic, but this implementation provides an initial step in that direction.

### CODE REVIEW

**Comment by fps7806** [comment]
Closed in favor of https://github.com/uber/petastorm/pull/540
